File: android2_lesson04/src/java/java/com/example/ama/android2_lesson04/ui/viewer/PVFragmentLoader.java
package com.example.ama.android2_lesson04.ui.viewer;

import android.graphics.Bitmap;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.Loader;
import android.view.View;

import com.example.ama.android2_lesson04.ServiceTestApp;
import com.example.ama.android2_lesson04.background.loader.MyLoader;
import com.example.ama.android2_lesson04.ui.viewer.base.BasePictureViewerFragment;
import com.example.ama.android2_lesson04.utils.ResourcesUtils;

import java.util.ArrayList;

public class PVFragmentLoader
        extends BasePictureViewerFragment
        implements LoaderManager.LoaderCallbacks<ArrayList<Bitmap>> {

    public static final int LOADER_ID = 1;

    public static PVFragmentLoader newInstance() {
        return new PVFragmentLoader();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        if (isLoading) {
            initLoader();
        }
    }

    @Override
    protected void onLoadImagesClick() {
        initLoader();
    }

    @NonNull
    @Override
    public Loader<ArrayList<Bitmap>> onCreateLoader(int id, @Nullable Bundle args) {
        return new MyLoader(ServiceTestApp.getInstance(), args);
    }

    @Override
    public void onLoadFinished(@NonNull Loader<ArrayList<Bitmap>> loader, ArrayList<Bitmap> data) {
        ServiceTestApp.getData().addBitmaps(data);
        showPictures(ServiceTestApp.getData().getBitmaps());
        onFinishLoading();
    }

    @Override
    public void onLoaderReset(@NonNull Loader loader) {

    }

    private void initLoader() {
        Bundle args = new Bundle();
        args.putStringArray(MyLoader.LOADER_ARGS, ResourcesUtils.getUrlsArray());
        if (getLoaderManager().getLoader(LOADER_ID) == null) {
            getLoaderManager().initLoader(LOADER_ID, args, this);
        } else {
            getLoaderManager().restartLoader(LOADER_ID, args, this);
        }
    }
}

File: android2_lesson04/src/kotlin/java/com/example/ama/android2_lesson04/ui/viewer/PVFragmentLoader.kt
package com.example.ama.android2_lesson04.ui.viewer

import android.graphics.Bitmap
import android.os.Bundle
import android.support.v4.app.LoaderManager
import android.support.v4.content.Loader
import android.view.View
import com.example.ama.android2_lesson04.R
import com.example.ama.android2_lesson04.ServiceTestApp
import com.example.ama.android2_lesson04.background.loader.LOADER_ARGS
import com.example.ama.android2_lesson04.background.loader.PictureLoader
import com.example.ama.android2_lesson04.ui.viewer.base.BasePictureViewerFragment

class PVFragmentLoader
    : BasePictureViewerFragment(),
        LoaderManager.LoaderCallbacks<ArrayList<Bitmap>> {

    companion object {
        const val LOADER_ID = 1
        fun newInstance() = PVFragmentLoader()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        if (isLoading) initLoader()
    }

    override fun onLoadImagesClick() {
        initLoader()
    }

    private fun initLoader() {
        val args = Bundle()
        args.putStringArray(LOADER_ARGS, resources.getStringArray(R.array.image_urls))
        if (loaderManager.getLoader<PictureLoader>(LOADER_ID) == null) {
            loaderManager.initLoader(LOADER_ID, args, this)
        } else {
            loaderManager.restartLoader(LOADER_ID, args, this)
        }
    }

    override fun onCreateLoader(id: Int, args: Bundle?) =
            PictureLoader(ServiceTestApp.instance, args)

    override fun onLoadFinished(loader: Loader<ArrayList<Bitmap>>, data: ArrayList<Bitmap>?) {
        if (data != null) {
            ServiceTestApp.data.bitmaps.addAll(data)
            showBitmaps(ServiceTestApp.data.bitmaps)
        }
        onFinishLoading()
    }

    override fun onLoaderReset(loader: Loader<ArrayList<Bitmap>>) {

    }

}